mongodb/src/main/java/org/hibernate/ogm/datastore/mongodb/MongoDBDialect.java

```diff
@@ -897,13 +896,13 @@ public int executeBackendUpdateQuery(final BackendQuery<MongoDBQueryDescriptor>
 	@Override
 	public MongoDBQueryDescriptor parseNativeQuery(String nativeQuery) {
 		NativeQueryParser parser = Parboiled.createParser( NativeQueryParser.class );
```

Comment:
This is preexisting to your patch but seeing this, I think we have made a design error here. The parser should be created only once and should not contain any state. The state should be kept in the value stack.

I think it's important to fix this now as you are adding a new state in your patch (`cliQueryInfo`) so it might be a bit harder to migrate to a value stack after these changes.

Reply:
I think it is right to use the value stack instead of keeping the state in the parser. I'll try to reimplement it this way.  I'm not sure though about using the state in the value stack because the parser run result is the state (builder) rather then the object we need (descriptor).  For me it feels awkward to use `run.resultValue.build()` when it could be simply `run.resultValue` and only the parser holds the errors and not both the builder and parser.  

If I understand Parboiled well enough, we could parameterize the parser over `Object` thus making the stack to have values of `Object`.  This way we don't really have to keep any state: the "lower level" rules can put information like the collection name to the stack and the "higher level" rules can pop it and make a decision to fail and put some errors in the error list (although then we have to keep casting things).